tools/lsp/properties.rs

```diff
@@ -121,13 +121,90 @@ fn add_element_properties(
     }))
 }
 
-fn left_extend(token: rowan::SyntaxToken): TextSize {
-    let start_position = token.text_range().start();
-    let mut current_token = token.previous_token();
+// Move left to include WS and comments that go with a token.
+fn left_extend(token: rowan::SyntaxToken<Language>) -> rowan::TextSize {
+    let expression_position = token.text_range().start();
+    let start_token = {
+        let mut current_token = token.prev_token();
+        let mut start_token = token.clone();
 
-    while (current_token) {
-        let current_token.previous_token();
+        // Walk backwards:
+        while let Some(t) = current_token {
+            if t.kind() != SyntaxKind::Whitespace && t.kind() != SyntaxKind::Comment {
+                break;
+            }
+            start_token = t.clone();
+            current_token = t.prev_token();
+        }
+        start_token
+    };
+
+    let start_position = start_token.text_range().start();
+    let len = expression_position
+        .checked_sub(start_position)
+        .expect("start_position >= start of start_token");
+    if len == 0.into() {
+        return start_position;
+    }
+
+    // Collect whitespace/comments into a string:
+    let possible_preamble = {
+        let mut tmp = String::with_capacity(len.into());
+        let mut current_token = start_token.clone();
+        while current_token.text_range().start() != expression_position {
```

Comment:
I wonder if it would not be simpler to merge all these loop in one. Just walk backward in the token, and stop when you get a whitespace token with a newline or a comment whose previous token doesn't have enough indentation.

Another possibility would be to make the parser smart enough to attach the comment to the node they belong.

But anyway, in a first iteration of this, let's not care too much about indentation or formatting, I'd say

Reply:
> I doubt that is really relevant anywhere else but in the LSP

Also usefull for the code formater and syntax updated, but arguably they should be in the LSP

> Why slow down the parser for things only the LSP can use?

Yes, that's a choice we make, we keep the whitespace and comments in the tree, we keep the nodes and such in the tree.
This shouldn't slow down the parser much anyway  (and we could have it behind a feature flag or so if we wanted)